File: frontend/src/types/salary.ts
/**
 * 급여 계산 API 타입 정의
 * 백엔드 API 스키마와 동기화
 */

export type EmploymentType = 'FULL_TIME' | 'PART_TIME';
export type CompanySize = 'UNDER_5' | 'OVER_5';
export type WageType =
  | 'MONTHLY'               // 하위 호환 → MONTHLY_FIXED
  | 'HOURLY'                // 하위 호환 → HOURLY_MONTHLY
  | 'MONTHLY_FIXED'         // 월급제 고정
  | 'HOURLY_MONTHLY'        // 시급제 월정산
  | 'HOURLY_BASED_MONTHLY'; // 시급기반 월급제
export type AbsencePolicy = 'STRICT' | 'MODERATE' | 'LENIENT';
export type HoursMode = '174' | '209';

/**
 * 4대 보험 적용 옵션
 */
export interface InsuranceOptions {
  apply_national_pension: boolean;    // 국민연금 적용
  apply_health_insurance: boolean;    // 건강보험 적용
  apply_long_term_care: boolean;      // 장기요양보험 적용
  apply_employment_insurance: boolean; // 고용보험 적용
}

export const DEFAULT_INSURANCE_OPTIONS: InsuranceOptions = {
  apply_national_pension: true,
  apply_health_insurance: true,
  apply_long_term_care: true,
  apply_employment_insurance: true,
};

/**
 * 포괄임금제 옵션
 */
export interface InclusiveWageOptions {
  enabled: boolean;                      // 포괄임금제 적용 여부
  fixed_overtime_hourly_rate: number;    // 연장수당 시간당 고정 금액 (원)
  monthly_expected_overtime_hours: number; // 월 예정 연장근로시간
}

export const DEFAULT_INCLUSIVE_WAGE_OPTIONS: InclusiveWageOptions = {
  enabled: false,
  fixed_overtime_hourly_rate: 0,
  monthly_expected_overtime_hours: 0,
};

export interface MoneyResponse {
  amount: number;
  formatted: string;
}

export interface WorkingHoursResponse {
  hours: number;
  minutes: number;
  total_minutes: number;
  formatted: string;
}

export interface EmployeeRequest {
  name: string;
  dependents_count: number;
  children_under_20: number;
  employment_type: EmploymentType;
  company_size: CompanySize;
  scheduled_work_days: number; // 주 소정근로일 (계약상 주당 근무일 수)
  daily_work_hours: number; // 1일 소정근로시간 (기본 8)
}

export interface AllowanceRequest {
  name: string;
  amount: number;
  is_taxable: boolean;
  is_includable_in_minimum_wage: boolean;
  is_fixed: boolean;
  is_included_in_regular_wage: boolean;
}

export interface WorkShiftRequest {
  date: string; // YYYY-MM-DD
  start_time: string; // HH:MM
  end_time: string; // HH:MM
  break_minutes: number;
  is_holiday_work: boolean;
}

export interface OvertimeBreakdown {
  overtime_hours: WorkingHoursResponse;
  overtime_pay: MoneyResponse;
  night_hours: WorkingHoursResponse;
  night_pay: MoneyResponse;
  holiday_hours: WorkingHoursResponse;
  holiday_pay: MoneyResponse;
  total: MoneyResponse;
}

export interface WeeklyHolidayPayBreakdown {
  amount: MoneyResponse;
  weekly_hours: WorkingHoursResponse;
  is_proportional: boolean;
  calculation: string;
}

export interface GrossBreakdown {
  base_salary: MoneyResponse;
  regular_wage: MoneyResponse;
  hourly_wage: MoneyResponse;
  taxable_allowances: MoneyResponse;
  non_taxable_allowances: MoneyResponse;
  overtime_allowances: OvertimeBreakdown;
  weekly_holiday_pay: WeeklyHolidayPayBreakdown;
  inclusive_overtime_pay?: MoneyResponse; // 포괄임금제 고정 연장수당
  total: MoneyResponse;
}

export interface InsuranceBreakdown {
  national_pension: MoneyResponse;
  health_insurance: MoneyResponse;
  long_term_care: MoneyResponse;
  employment_insurance: MoneyResponse;
  total: MoneyResponse;
  applied_options?: InsuranceOptions;
}

export interface TaxBreakdown {
  income_tax: MoneyResponse;
  local_income_tax: MoneyResponse;
  total: MoneyResponse;
}

export interface DeductionsBreakdown {
  insurance: InsuranceBreakdown;
  tax: TaxBreakdown;
  total: MoneyResponse;
}

export interface SalaryCalculationRequest {
  employee: EmployeeRequest;
  base_salary: number;
  allowances: AllowanceRequest[];
  work_shifts: WorkShiftRequest[];
  wage_type: WageType;
  hourly_wage: number;
  calculation_month: string;
  absence_policy: AbsencePolicy;
  hours_mode: HoursMode;
  insurance_options?: InsuranceOptions;
  weekly_hours?: number; // 주 근무시간 (기본 40)
  inclusive_wage_options?: InclusiveWageOptions; // 포괄임금제 옵션
  contract_monthly_salary?: number; // 시급기반 월급제 계약 월급
}

export interface WorkSummaryResponse {
  calculation_month: string;
  wage_type: string;
  scheduled_days: number;
  actual_work_days: number;
  absent_days: number;
  total_work_hours: WorkingHoursResponse;
  regular_hours: WorkingHoursResponse;
  overtime_hours: WorkingHoursResponse;
  night_hours: WorkingHoursResponse;
  holiday_hours: WorkingHoursResponse;
  weekly_holiday_weeks: number;
  total_weeks: number;
}

export interface AbsenceBreakdown {
  scheduled_days: number;
  actual_work_days: number;
  absent_days: number;
  daily_wage: MoneyResponse;
  wage_deduction: MoneyResponse;
  holiday_pay_loss: MoneyResponse;
  total_deduction: MoneyResponse;
  absence_policy: string;
}

export interface WarningItem {
  level: 'critical' | 'warning' | 'info';
  message: string;
  detail: string;
}

export interface InclusiveWageOptionsResponse {
  enabled: boolean;
  fixed_overtime_hourly_rate: number;
  monthly_expected_overtime_hours: number;
  monthly_fixed_overtime_pay?: MoneyResponse;
}

export interface SalaryCalculationResponse {
  employee_name: string;
  gross_breakdown: GrossBreakdown;
  deductions_breakdown: DeductionsBreakdown;
  net_pay: MoneyResponse;
  work_summary?: WorkSummaryResponse;
  absence_breakdown?: AbsenceBreakdown;
  inclusive_wage_options?: InclusiveWageOptionsResponse;
  applied_wage_mode?: 'CONTRACT_SALARY' | 'ACTUAL_CALCULATION';
  contract_vs_actual_diff?: MoneyResponse;
  warnings: WarningItem[];
  calculation_metadata: {
    calculation_date: string;
    tax_year: number;
    insurance_year: number;
    wage_type?: string;
  };
}

export interface ReverseSalaryRequest {
  target_net_pay: number;
  employee: EmployeeRequest;
  allowances: AllowanceRequest[];
  work_shifts: WorkShiftRequest[];
  wage_type: WageType;
  calculation_month: string;
  absence_policy: AbsencePolicy;
}

export interface ReverseSalaryResponse {
  target_net_pay: MoneyResponse;
  required_base_salary: MoneyResponse;
  actual_net_pay: MoneyResponse;
  difference: MoneyResponse;
  iterations: number;
  calculation_result: SalaryCalculationResponse;
  warnings: WarningItem[];
}

export interface ErrorResponse {
  error: string;
  detail: string;
}

File: backend/app/api/schemas/salary.py
"""급여 계산 스키마"""
from typing import List, Dict, Any, Literal, Optional
from pydantic import BaseModel, Field, ConfigDict

from .common import (
    EmployeeRequest,
    AllowanceRequest,
    WorkShiftRequest,
    MoneyResponse,
    WorkingHoursResponse,
)


class SalaryCalculationRequest(BaseModel):
    """급여 계산 요청"""
    employee: EmployeeRequest = Field(..., description="근로자 정보")
    base_salary: int = Field(0, ge=0, description="기본 월급 (월급제)")
    allowances: List[AllowanceRequest] = Field(
        default_factory=list, description="수당 목록"
    )
    work_shifts: List[WorkShiftRequest] = Field(
        default_factory=list, description="근무 시프트 목록"
    )
    wage_type: Literal["MONTHLY", "HOURLY"] = Field(
        "MONTHLY", description="급여 형태 (MONTHLY=월급제, HOURLY=시급제)"
    )
    hourly_wage: int = Field(
        0, ge=0, description="시급 (시급제일 때 사용)"
    )
    calculation_month: str = Field(
        "", description="계산 대상 월 (YYYY-MM, 빈 값이면 시프트 날짜에서 추론)"
    )
    absence_policy: Literal["STRICT", "MODERATE", "LENIENT"] = Field(
        "STRICT", description="결근 공제 정책 (월급제 전용)"
    )
    hours_mode: Literal["174", "209"] = Field(
        "174", description="월 소정근로시간 계산 방식 (174=주휴분리, 209=주휴포함)"
    )

    model_config = ConfigDict(json_schema_extra={
        "example": {
            "employee": {
                "name": "홍길동",
                "dependents_count": 2,
                "children_under_20": 1,
                "employment_type": "FULL_TIME",
                "company_size": "OVER_5"
            },
            "base_salary": 2500000,
            "allowances": [
                {
                    "name": "직책수당",
                    "amount": 300000,
                    "is_taxable": True,
                    "is_includable_in_minimum_wage": True,
                    "is_fixed": True,
                    "is_included_in_regular_wage": True
                }
            ],
            "work_shifts": [
                {
                    "date": "2026-01-05",
                    "start_time": "09:00:00",
                    "end_time": "18:00:00",
                    "break_minutes": 60,
                    "is_holiday_work": False
                }
            ]
        }
    })


class InsuranceBreakdown(BaseModel):
    """4대 보험 상세"""
    national_pension: MoneyResponse = Field(..., description="국민연금")
    health_insurance: MoneyResponse = Field(..., description="건강보험")
    long_term_care: MoneyResponse = Field(..., description="장기요양보험")
    employment_insurance: MoneyResponse = Field(..., description="고용보험")
    total: MoneyResponse = Field(..., description="4대 보험 합계")


class TaxBreakdown(BaseModel):
    """세금 상세"""
    income_tax: MoneyResponse = Field(..., description="소득세")
    local_income_tax: MoneyResponse = Field(..., description="지방소득세")
    total: MoneyResponse = Field(..., description="세금 합계")


class OvertimeBreakdown(BaseModel):
    """가산수당 상세"""
    overtime_hours: WorkingHoursResponse = Field(..., description="연장근로시간")
    overtime_pay: MoneyResponse = Field(..., description="연장근로수당")
    night_hours: WorkingHoursResponse = Field(..., description="야간근로시간")
    night_pay: MoneyResponse = Field(..., description="야간근로수당")
    holiday_hours: WorkingHoursResponse = Field(..., description="휴일근로시간")
    holiday_pay: MoneyResponse = Field(..., description="휴일근로수당")
    total: MoneyResponse = Field(..., description="가산수당 합계")


class WeeklyHolidayPayBreakdown(BaseModel):
    """주휴수당 상세"""
    amount: MoneyResponse = Field(..., description="주휴수당")
    weekly_hours: WorkingHoursResponse = Field(..., description="주 평균 근로시간")
    is_proportional: bool = Field(..., description="비례 지급 여부")
    calculation: str = Field(..., description="계산식")


class GrossBreakdown(BaseModel):
    """총 지급액 상세"""
    base_salary: MoneyResponse = Field(..., description="기본급")
    regular_wage: MoneyResponse = Field(..., description="통상임금")
    hourly_wage: MoneyResponse = Field(..., description="통상시급")
    taxable_allowances: MoneyResponse = Field(..., description="과세 수당")
    non_taxable_allowances: MoneyResponse = Field(..., description="비과세 수당")
    overtime_allowances: OvertimeBreakdown = Field(..., description="가산수당")
    weekly_holiday_pay: WeeklyHolidayPayBreakdown = Field(
        ..., description="주휴수당"
    )
    total: MoneyResponse = Field(..., description="총 지급액")


class DeductionsBreakdown(BaseModel):
    """공제 내역 상세"""
    insurance: InsuranceBreakdown = Field(..., description="4대 보험")
    tax: TaxBreakdown = Field(..., description="세금")
    total: MoneyResponse = Field(..., description="총 공제액")


class AbsenceBreakdown(BaseModel):
    """결근 공제 상세 (월급제 전용)"""
    scheduled_days: int = Field(..., description="소정근로일수")
    actual_work_days: int = Field(..., description="실제 근무일수")
    absent_days: int = Field(..., description="결근일수")
    daily_wage: MoneyResponse = Field(..., description="일급")
    wage_deduction: MoneyResponse = Field(..., description="일급 공제액")
    holiday_pay_loss: MoneyResponse = Field(..., description="주휴수당 미지급액")
    total_deduction: MoneyResponse = Field(..., description="총 결근 공제액")
    absence_policy: str = Field(..., description="적용된 공제 정책")


class WorkSummaryResponse(BaseModel):
    """근무 요약"""
    calculation_month: str = Field(..., description="계산 대상 월")
    wage_type: str = Field(..., description="급여 형태")
    scheduled_days: int = Field(..., description="소정근로일수")
    actual_work_days: int = Field(..., description="실제 근무일수")
    absent_days: int = Field(..., description="결근일수")
    total_work_hours: WorkingHoursResponse = Field(..., description="총 근무시간")
    regular_hours: WorkingHoursResponse = Field(..., description="소정근로시간")
    overtime_hours: WorkingHoursResponse = Field(..., description="연장근로시간")
    night_hours: WorkingHoursResponse = Field(..., description="야간근로시간")
    holiday_hours: WorkingHoursResponse = Field(..., description="휴일근로시간")
    weekly_holiday_weeks: int = Field(..., description="주휴수당 발생 주 수")
    total_weeks: int = Field(..., description="해당 월 총 주 수")


class WarningResponse(BaseModel):
    """경고 메시지"""
    level: str = Field(..., description="경고 수준 (critical, warning, info)")
    message: str = Field(..., description="경고 메시지")
    detail: str = Field(default="", description="상세 설명")


class ReverseSalaryRequest(BaseModel):
    """역산 계산 요청 (실수령액 → 필요 기본급)"""
    target_net_pay: int = Field(..., gt=0, description="목표 실수령액")
    employee: EmployeeRequest = Field(..., description="근로자 정보")
    allowances: List[AllowanceRequest] = Field(
        default_factory=list, description="수당 목록"
    )
    work_shifts: List[WorkShiftRequest] = Field(
        default_factory=list, description="근무 시프트 목록"
    )
    wage_type: Literal["MONTHLY", "HOURLY"] = Field(
        "MONTHLY", description="급여 형태"
    )
    calculation_month: str = Field(
        "", description="계산 대상 월 (YYYY-MM)"
    )
    absence_policy: Literal["STRICT", "MODERATE", "LENIENT"] = Field(
        "STRICT", description="결근 공제 정책"
    )


class ReverseSalaryResponse(BaseModel):
    """역산 계산 응답"""
    target_net_pay: MoneyResponse = Field(..., description="목표 실수령액")
    required_base_salary: MoneyResponse = Field(..., description="필요 기본급")
    actual_net_pay: MoneyResponse = Field(..., description="실제 계산된 실수령액")
    difference: MoneyResponse = Field(..., description="오차 (±1,000원 이내)")
    iterations: int = Field(..., description="탐색 반복 횟수")
    calculation_result: 'SalaryCalculationResponse' = Field(
        ..., description="상세 급여 계산 결과"
    )
    warnings: List[WarningResponse] = Field(
        default_factory=list, description="경고 메시지"
    )


class SalaryCalculationResponse(BaseModel):
    """급여 계산 응답"""
    employee_name: str = Field(..., description="근로자 이름")
    gross_breakdown: GrossBreakdown = Field(..., description="지급 내역")
    deductions_breakdown: DeductionsBreakdown = Field(..., description="공제 내역")
    net_pay: MoneyResponse = Field(..., description="실수령액")
    work_summary: Optional[WorkSummaryResponse] = Field(None, description="근무 요약")
    absence_breakdown: Optional[AbsenceBreakdown] = Field(None, description="결근 공제 상세")
    warnings: List[WarningResponse] = Field(
        default_factory=list, description="경고 메시지 목록"
    )
    calculation_metadata: Dict[str, Any] = Field(
        default_factory=dict, description="계산 메타데이터"
    )

    model_config = ConfigDict(json_schema_extra={
        "example": {
            "employee_name": "홍길동",
            "gross_breakdown": {
                "base_salary": {"amount": 2500000, "formatted": "2,500,000원"},
                "total": {"amount": 3559358, "formatted": "3,559,358원"}
            },
            "deductions_breakdown": {
                "insurance": {
                    "total": {"amount": 317814, "formatted": "317,814원"}
                },
                "tax": {
                    "total": {"amount": 48345, "formatted": "48,345원"}
                },
                "total": {"amount": 366159, "formatted": "366,159원"}
            },
            "net_pay": {"amount": 3193199, "formatted": "3,193,199원"},
            "warnings": [],
            "calculation_metadata": {
                "calculation_date": "2026-01-13",
                "tax_year": 2026,
                "insurance_year": 2026
            }
        }
    })
